### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by 00jonghyeon00**
Building a neural network_numpy

### CODE REVIEW

File: Numpy_Building a neural network.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import pandas as pd
```
Comment (kypark-hs) [inline Numpy_Building a neural network.py:1]: python script 최상단에 이런 선언으로 시작하시길 추천드립니다:

```
#!/usr/bin/env python
# -*- coding:utf-8 -*-
```
무슨 의미인지 모른다면 꼭 찾아보세요.

File: Numpy_Building a neural network.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import numpy as np
```
Comment (kypark-hs) [inline Numpy_Building a neural network.py:2]: linting tool을 사용하시기를 추천드립니다. 대표적인 예로 pylint가 있습니다.
아울러 package import 순서도 가급적 원칙을 정해두고 지키시기 바래요.
1차적으로는 linting tool의 가이드라인을 따르시고 그 외의 경우엔 alphabetical order를 지킨다든지...
coding convention에서 강제하거나 linting tool 등의 수단이 적용되지 않는 경우, 개인적인 코딩 룰을 정하시고 지키세요.

File: Numpy_Building a neural network.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    return W1, b1, W2, b2
+
+def ReLU(Z):
+    return np.maximum(0, Z)
+
+def ReLU_deriv(Z):
+    return Z > 0
+
+def softmax(Z):
+    expZ = np.exp(Z - np.max(Z, axis=0, keepdims=True))
```
Comment (kypark-hs) [inline Numpy_Building a neural network.py:33]: 본인의 로직인가요? np.exp(Z)를 사용하지 않은 이유는 무엇인가요?

File: Numpy_Building a neural network.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        Z1, A1, Z2, A2 = forward_propagation(W1, b1, W2, b2, X)
+        dW1, db1, dW2, db2 = backpropagation(Z1, A1, Z2, A2, W1, W2, X, Y)
+        W1, b1, W2, b2 = update_params(W1, b1, W2, b2, dW1, db1, dW2, db2, LR)
+        if i % 10 == 0:
+            print("Iter:", i)
+            preds = predictions(A2)
+            print("Accuracy:", accuracy(preds, Y))
+    return W1, b1, W2, b2
+
+W1, b1, W2, b2 = gradient_descent(X_train, Y_train, 0.001, 300)
```
Comment (kypark-hs) [inline Numpy_Building a neural network.py:85]: 기왕 하는 김에 X_dev, Y_dev로 inference도 해보시죠 ㅎㅎ

File: Numpy_Building a neural network.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+def gradient_descent(X, Y, LR, iterations):
+    W1, b1, W2, b2 = init_params()
+
+    for i in range(iterations):
+        Z1, A1, Z2, A2 = forward_propagation(W1, b1, W2, b2, X)
+        dW1, db1, dW2, db2 = backpropagation(Z1, A1, Z2, A2, W1, W2, X, Y)
+        W1, b1, W2, b2 = update_params(W1, b1, W2, b2, dW1, db1, dW2, db2, LR)
+        if i % 10 == 0:
+            print("Iter:", i)
+            preds = predictions(A2)
```
Comment (kypark-hs) [inline Numpy_Building a neural network.py:81]: 가급적 loss 값 계산도 추가해 보시기를 추천 드립니다. 대개는 cross entropy를 사용하죠...

**Review by kypark-hs — COMMENTED** [review]
전반적으로 잘 하셨습니다.
본인 스스로 흥미가 있다면 초기화 부분이나 activation function을
다양하게 실험해 보시는것도 추천드립니다.

File: Numpy_Building a neural network.py (hunk 3 above)
Comment (00jonghyeon00) [reply Numpy_Building a neural network.py:33]: 오버플로우랑 언더플로우 고려해야한다는 글을 본적이 있어서 이렇게 했습니다.

File: Numpy_Building a neural network.py (hunk 1 above)
Comment (00jonghyeon00) [reply Numpy_Building a neural network.py:1]: 1. `#!/usr/bin/env python `
 - 'shebang' or 'hanhbang' 으로 불리며 script를 실행할 때 사용할 interpreter를 지정

2. `# -*- coding:utf-8 -*-`
 - 문자 인코딩 지정(Python3에서는 기본 인코딩이 UTF-8)

File: Numpy_Building a neural network.py (hunk 3 above)
Comment (kypark-hs) [reply Numpy_Building a neural network.py:33]: softmax layer의 입력인 logit 실수 값들이 커질 경우 np.exp(exponentiate) 함수 사용 시 무한대 수준으로 발산할 수 있다는
위험 때문에 정규화 작업을 적용하기는 합니다. 다만 overflow를 막기 위해 X - np.max(Z)를 한다면 거꾸로 underflow 문제가
생기게 되지 않나요? ㅎ
normalization 혹은 regularization(둘을 한국어로 흔히 '정규화'라고 동일하게 번역하는 경우들이 많은데 엄연히 다른 개념)을
적용할 때 잘 따져보고 해야 합니다.
overflow/underflow 같은 절대값 크기 수준의 문제에 대응하는 방법으로는 clipping 같은 작업을 더 많이 사용하기는 합니다...
어떤 작업을 하는 것인지 제대로 이해하고 쓰는 것인가 싶어서 물어봤습니다. ㅎㅎ

File: Numpy_Building a neural network.py (hunk 1 above)
Comment (kypark-hs) [reply Numpy_Building a neural network.py:1]: 추가로 의견을 드리자면...
python의 경우 요즘은 거의(무조건) python3를 기본으로 간주하기 때문에 아예 문제를 뭉개버리는 상황이긴 합니다만...
얼마전까지만 해도 python2와 python3간의 하위 호환성 문제가 저 바닥에서는 무척 큰 골치거리였습니다.

`# -*- coding:utf-8 -*-`

선언은 python2와의 하위 호환성을 고려해서 스크립트 머릿말로 습관적으로 추가하곤 했던 선언문입니다. 요즘엔 그냥 python3를
다 쓴다고 간주해버리기 때문에 문제를 그냥 무시하는 경향이 있지만 엄밀히 말하자면 하위 호환성 문제가 존재하는 실수입니다.
가급적이면 선언하고 사용하시기 바래요...

Comment (00jonghyeon00) [reply Numpy_Building a neural network.py:1]: 넵! 피드백 감사합니다!
